pilz_control/test/integrationtest_pilz_joint_trajectory_controller.py

```diff
@@ -121,155 +120,197 @@ def observeStop(self, max_deceleration = MAX_DECELERATION, timeout = STOP_DURATI
-        trajectory_dispatcher.dispatchTrajectory(goal_position = 0.2, time_from_start = 5)
+    def setUp(self):
+        self.assertTrue(self._turn_off_speed_monitoring(), 'Could not turn off speed monitoring')
+
+    def _move_to(self, position, duration, expected_error_code=FollowJointTrajectoryResult.SUCCESSFUL):
+        self._start_motion(position=position, duration=duration)
+        return self._wait_for_motion_result(expected_error_code)
+
+    def _start_motion(self, position, duration):
+        self._trajectory_dispatcher.dispatch_trajectory(goal_position=position, time_from_start=duration)
+
+    def _wait_for_motion_result(self, expected_error_code=FollowJointTrajectoryResult.SUCCESSFUL):
+        result = self._trajectory_dispatcher.wait_for_result()
+        return result.error_code == expected_error_code
+
+    def _turn_off_speed_monitoring(self):
+        return self._monitored_cartesian_speed_srv.turn_off_speed_monitoring()
+
+    def _turn_on_speed_monitoring(self):
+        return self._monitored_cartesian_speed_srv.turn_on_speed_monitoring()
+
+    def test_hold_at_startup(self):
+        """Tests that 'hold' mode is active at startup of controller."""
+
+        self.assertTrue(self._move_to(position=DEFAULT_GOAL_POSITION, duration=1,
+                                      expected_error_code=FollowJointTrajectoryResult.INVALID_GOAL),
+                        "Motion did not fail although controller should be in 'hold' mode at startup")
+
+        self.assertTrue(self._hold_srv.request_default_mode(), 'Switch to default mode failed')
+
+        self.assertTrue(self._move_to(position=DEFAULT_GOAL_POSITION, duration=0.5),
+                        "Motion failed although controller should be in 'unhold' mode")
+
+    def test_hold_during_motion(self):
+        """Activate hold mode while robot is moving and evaluate executed stop."""
+        is_executing_srv = IsExecutingServiceWrapper()
+
+        self._start_motion(position = 0.2, duration = 5)
 
         # Wait for movement to commence
-        MovementObserver().observeUntilPositionGreater(0.11, 3)
-        self.assertTrue(is_executing_srv.call(), 'is_executing service returned false.')
-        # Hold robot during movement
-        rospy.loginfo("  - Robot is moving -> Switch into HOLDING mode")
-        self.assertTrue(hold_srv.requestHoldingMode(), 'Switch to Mode HOLDING failed.')
+        MovementObserver().observe_until_position_greater(0.11, 3)
+        self.assertTrue(is_executing_srv.call(), "Controller is not executing")
+        rospy.loginfo("Robot is moving -> Switch into HOLDING mode")
+        self.assertTrue(self._hold_srv.request_holding_mode(), 'Switch to Mode HOLDING failed.')
 
         # Make sure robot stops (not abruptly)
-        MovementObserver().observeStop()
-        rospy.loginfo("  - Stop of robot observed!")
+        MovementObserver().observe_stop()
```

Comment:
I know this existed before, but ...

The robot stops already whilst the hold request is processed. So this check has to be done in parallel to the hold request, else it is meaningless.

Reply:
OK, if the robot really already stops while the hold mode is requested, I suggest to make the motion "longer". Because everything already happens in parallel. The observer simply waits until we are sure that the motion is in progress. The motion itself does not happen in the Test-Framework but in the ROS system in parallel. If we are sure that a motion happens, we trigger the hold. A thread does not improve this situation, only a longer motion, so that we have enough time to react.

Reply:
Done.